### ISSUE TITLE
Integration testing / Connected to CRM features

### ISSUE THREAD

**Issue by jordimontana82** [jordimontana82/fake-xrm-easy#87]
Hi guys,

Adding this based on feedback of adding the possibility of running tests against real CRM organisation instances.

Feedback received so far involves 2 different approaches:

- 1)  Add integration testing to the existing NuGet package

So that one could run tests against a real CRM org by using an app.config parameter to store the connection string and an appSettings key which would be the execution mode (Unit / Integration). 

With this approach there will be a single test project where tests would use a fake service or a real service based on the appSetting key (which could be dynamically changed via the Configuration settings in VS).

- 2) Add integration testing to a brand new NuGet package

With this approach, there would be 2 Test projects in the solution, one for unit testing and another one for integration testing. The integration testing package would be really simple, because it doesn't have to use any fakes, it would a real CRM organisation service instance + some helper methods really. It would store the connection string in the app.config as well, but in this case there is no need for an app setting to know if we are doing unit test or integration testing as this will be implicit by the fact of using a different NuGet package.

Regardless of the approach chosen, there are big differences about how you would test code between unit & integration testing.

For unit testing, you want to execute a plugin directly, but for integration testing, that plugin will be actually triggered via a CRM message (i.e. Create / Update / etc...).

So, in order to trigger the plugin, from a unit testing perspective one could do:

    ctx.ExecutePluginWith...

Whereas for integration testing, one will have to trigger the plugin via a message, like:

    service.Create(new Entity())

There are other complexities in integration testing, which is that we have asynchronous code (like, workflows, async plugins, etc). It would be nice to have helper methods in that scenario, whichwould wait until a workflow is completed for example, so that we can do Assertions after that.

Like:

    WaitForWorkflow(string sEntityName, Guid gRecordId, Action<SystemJob> lamba)

So that we can do:

    ctx.WaitForWorkflow("some entity name", {guid}, (SystemJob sj) => {
                                       Assert.True(whatever here....)
                              });

Example, say we have a workflow which creates a follow up email activity when some attribute of a contact record is updated.

We could "integrate test" that by doing:
   
    var contactId = new Guid("{hardcoded guid from your environment}");
    var contact = service.Retrieve("contact", contactId);
  
    //Number of email activities before test
    var numberOfEmailsBeforeTest = ctx.CreateQuery<Email>()
                                             .Where(e => e.RegardingObjectId.Id== contactId)
                                             .ToList()
                                             .Count();

    //Trigger an update in the contact record, this should, trigger a workflow at some stage
    service.Update(contact);

    //Wait for a workflow instance (system job) to be created for that contact record
    ctx.WaitForWorkflow("contact", contactId, (SystemJob sj) => {

                                      //Asserts here....

                                       Assert.Equal(sj.StatusReason, Status.Succeeded); //Job succeeded

                                       var numberOfEmailsAfterTest = ctx.CreateQuery<Email>()
                                             .Where(e => e.RegardingObjectId.Id== contactId)
                                             .ToList()
                                             .Count();
                                       
                                      //A new email was created with the contact as the regarding...
                                      Assert.Equal(numberOfEmailsAfterTest , numberOfEmailsBeforeTest + 1);
                              });

Personally, I think approach 2 is better, cause we can version / mantain the 2 packages separately so that changes for the unit test package won't necessarily affect the integration testing package and viceversa.

Any thoughs? Feel free to add anything (even completely different approaches / solutions for this if needed!)

**Comment by jordimontana82** [comment]
@DigitalFlow @bacon1986 @ccellar

What do you think guys?

**Comment by rmsryu** [comment]
Hi Jordi, 

The part where it's said.. "running tests against real CRM organisation" it's the part I'm interested in, which is option #1. But having another one to ensure functionality with real data "integration test", I think this could be interesting as well.. Do I need to choose? or can i have both? :)

About the option #1. I will explain which are here my needs.. And I can't call it integration test or unit test. It's just bug fixing work!! But here it goes. Sometimes you have the need to debug a plugin or workflow, my unit tests are all in green but it's not working in the server and CRM logs doesn't helps at all. I might have forgot something really stupid.. Or someone else has just added functionality which breaks my logic. So what can I do? 
Use pluginregistration tool.... It's just so... "I'm too lazy"...

> Find PluginRegistration tool... I can't find this program :)
> Connect to your organisation.
> Wait for discover things....
> Choose your assembly..
> Install profiler.. 
> Here hangs for a minute... or two... 
> Execute the action in crm... If you still can remember what you are doing...
> Export to error to local file 
> Open Debugger...... 
> buffff...
> Select your error.txt file... 
> DON'T FORGET TO ATTACH VS... 
> And now run. 
> Did you didn't find out what you were looking for? NO start again... AAAGGHHH!!

I think you get the idea... Why can I just not run my plugin/workflow against CRM? If I have unit tests done I could reuse one of them, remove the bits of the in memory data and just execute and debug against one record on the server.

As mentioned, this is for bug fixing work. So it won't fit with Unit Test or Integration Test. But it would be really helpful in some situations.

Thanks
Raul

**Comment by jordimontana82** [comment]
Haha! Indeed, too many steps to remotely debug a plugin.....

So option 1 would do:

If configuration = "Connected" just use a real organisation service, and ignore the .Initialize() method.

I just can't think of a good name for this, cause when you do:

    var ctx = new XrmFakedContext()
    var service = ctx.GetFakedOrganizationService() // this won't be a fake service, but a real one :)

What name would you give to this? "Troubleshooting mode", "Firefighter mode"? : D

**Comment by craigseymour** [comment]
I'm with @rmsryu.  The reason I first really started to look at this type of framework was because I was debugging against live environments (using a console app to make the connection) and was wondering what the best way was to fake stuff up without having to make that live connection.

I can't help you with the naming, other than saying that my console application is called 'ConsoleApplicationForDebugging' ;-)

**Comment by jordimontana82** [comment]
: )

Right! I'd probably rename it to var service = ctx.GetOrganizationService(), which would retrieve a fake / real service based on configuration, and mark the existing method as "to be deprecated" in future.

Votes?

**Comment by craigseymour** [comment]
+1 - The Method is retrieving the org service for the context - I don't think I really care that it is fake or not - that is a result of the context I'm retrieving it from.  It should be transparent to me that it too is fake - because it's been written so awesomely and nothing goes wrong with it ;-)

**Comment by daryllabar** [comment]
Ideally I would love to abandon the [XrmUnitTest](https://github.com/daryllabar/XrmUnitTest library since I will never approach the level of documentation that this library has, and @jordimontana82, you have done a great job of getting a following of people supporting the framework.  I have none currently. :(

Currently I fully support the Option 1 Paradigm without having to create separate integration tests.  There are three "hoops" you have to jump through to support this though.

1.  Define all of your entity Ids that you will be creating as a part of your test ahead of time, so they can automagically be cleaned up pre and post test.
2.  Define any entities you are going to assume exist in the environment.  This allows for you to specify Global or default values, that are serialized when running against a real live CRM instance, so they can be used when running against a Fake instance.
3.  Handle any potential issues with existing plugins/workflows that may be manipulating the data.  Lets say you are testing a plugin that creates a child record.  In your test you assert that only one child record exists, after you run your test.  Well, if the plugin is registered against CRM, it will also create the child, and your test will fail, since there will be two child records, rather than one.  This normally just involves cleaning up some data after the data-initialization for the plugin, but before the actual plugin event.

**Comment by nicknow** [comment]
My thought is this should be an extension of FakeXrmEasy, not an addition. These capabilities, as valuable as they might be, are beyond standard coded Unit Testing and Integration testing. It is a new capability and should go to its own NuGet package and be treated a separate project. We would not want to get into a situation that FakeXrmEasy starts to become polluted by requirements that go beyond its core faking capabilities.

I find a lot of value in FakeXrmEasy for exactly what it is today. I get two significant benefits from using FakeXrmEasy.

1. I can create a faked `IOrganizationService` to test my business logic and query classes - this, the in-memory CRM service, is the most valuable. It has saved countless hours debugging queries. If you write all your queries in an isolated class you can write unit tests around just the queries. FakeXrmEasy makes it very easy to do these tests.
2.  I can test core plug-in execution (i.e., `IPlugin.Execute`.) I work hard to ensure that an IPlugin class only contains the logic required for the plugin itself, anything beyond analyzing `IPluginExecutionContext` to extract a Target and/or Images, check for a value change, etc. is handed off to a separate business logic class which is tested separately. And using the concept of dependency injection isolated from the `IPlugin` class.

So while I'm not against expanding the capabilities offered by `FakeXrmEasy` I don't want to solution become more focused on server integration testing - which is a complicated activity and will need a lot of support - at cost of continuing to enhance and make more robust the core CRM in-memory context.

What I'm thinking is that `FakeXrmEasy` is modified to have an interface `IXrmFakedContext` and `XrmFakedContext` implements the interface. Then in a new package, FakeXrmEasyReal` (just needed a name for this comment) package there would be a class, `XrmFakedContextWithRealSupport` (implementing `IXrmFakedContext`), that would return either an instance of `IXrmFakedContext FakeXrmEasy.XrmFakedContext()` _or_ `IXrmFakedContext FakeXrmEasyReal.XrmFakedContextWithRealSupport()`.

As to @daryllabar point # 2: One area of work that might be very valuable (and I may, if time permits, take a look at getting it going myself) is creating an enhancement and tool to make it easier to hydrate an instance of `XrmFakedContext`. Right now it is all coded. What I would like to see is an application that can read data from an existing Dynamics CRM instance, save that data (i.e., a JSON file), and then read them into a `List<Entity>` object to feed to `XrmFakedContext.Initialize`.

As to @daryllabar point # 3: Perhaps we should consider allowing the registration of plugins and workflows in an `XrmFakedContext`. These could either be in the project being tested or could be a DLL reference. Then we could have a method ` void WaitForAsyncPipelineCompletion()` to pause the test until an async plugins and/or workflows that were registered have fired. (This is not trivial because there is a lot of logic that goes along with maintaining the pipeline information correctly - is this depth 1 or 2 or 3, etc.)

These are just my initial thoughts and I reserve the right to change them - as I become convinced I'm wrong (or I come up with new ideas.)

**Comment by rmsryu** [comment]
The way I see "Integration Test", it's basically run some predefined scenarios in our integration test server, and evaluate the results.  I will also assume that all the code involved.. Plugins and workflows, has already been unit tested (100% of code coverage of course). 

So I would need I way to setup my scenarios (or test plan), and when I'm done, do a rollback (teardown) of my changes. However I will leave the setup and tear-down  to the testing framework, for instance in MSTest [TestCleanup] and [TestInitialize] attributes. 

In this situation. I will not have control of what's going on in my plugins, except we are attached to the process with the remote debugger (assuming we have access to the server). So I just need to know that if I do A B and C, I will get X Y and Z. And that's it. 

So I will need things like you have already described above, and with special consideration of the Async bits.

```

try
{
 //Validations Asserts.
 service.Create(record).ItShouldThrowAnException( e => 
 {
    Assert.True(e.Message, "AddresLine1 it's required.");
 });

 //Async Operations
 //Arranges
 context.SetTimeOut = 500; //Time I will wait for check again things in the server
 context.SetMaxTimeOut = 3000; //Maximum time we will wait until we throw timeoutexception.

 //Act 
 service.Update(entity);

 //Asserts
 Assert.True(context.WaitForWorkflow(workflowid).WithSuccess);
 Assert.True(context.WaitUntil(fetchXml).Returns.AtLeastOne);
 Assert.True(context.WaitForStatusChanged(myentity, State, Status));
 .. etc..

}catch(TimeOutException e){
  //Oopss!!
}
```

But I wouldn't use FakeXrmEasy for that I would use some set of extensions (brand new) of my crm context. I want FakeXrmEasy for unit test. Not for Integration test server side. I originally suggested to Jordi that it would be nice for troubleshooting, to quickly convert our FakeXrmContext into a real service context. So in a minute you would have your test up and running for debugging. But this can also be achieve with your ConsoleApplicationForDebugging @craigseymour :)

So in summary.. @jordimontana82  to your option 2 :+1:

**Comment by jordimontana82** [comment]
Hi guys, this has been an awesome feedback!

Thx @daryllabar, it's great that many people have contributed to the project. Special thanks to @DigitalFlow who sent a good amount of pull requests and wrote an amazing FAKE script (in F#) to version, build, run the tests, and deploy the 4 versions of the NuGet packages! I just extended it to run code coverage reports too.

There have been many posts so I'll try to recap.

Basically we need 2 different things (I should probably create 2 separate issues for each of them, actually):  

1)  Being able to quickly run existing unit tests against a real CRM organisation, just for the purpose of debugging production / real issues, and....
2)  Integration testing

**1) Troubleshooting existing code from unit tests, but against a real CRM service**

For this one, and after reading all the feedback, kind of agree to what both @nicknow and @rmsryu are proposing. We do have an [IXrmFakedContext interface already](https://github.com/jordimontana82/fake-xrm-easy/blob/7553d4697f1ce99d6ff0f7e663c60d47c9b9bed1/FakeXrmEasy.Shared/IXrmFakedContext.cs), so we could create a new class which implements that one for running tests against the real organisation service which could be called XrmRealContext, for instance.  

This class would implement the same interface and it will be really simple, not a big addition to framework because it just needs to return a CRM Organisation instance using the Xrm.Sdk and nothing else. 

We should probably rename it to something more generic, like IXrmContext, and the main method IXrmContext.GetFakedOrganizationService() to IXrmContext.GetOrganizationService() probably, cause the new interface will have 2 classes implementing it, the XrmFakedContext and the real one.

This would keep things seperated nicely and clean enough, and, when debugging a unit test against a real CRM organisation, we just need to change one class to another  XrmFakedContext -> XrmRealContext, which is just 1 line of code and it should be way more quicker than just debugging the code itself... ;)  

Once the issue has been fixed, and tested, we could just restore the class back to the fake context one XrmFakedContext and we are done.

I would add any other extensions / features for integration testing into another NuGet package, like FakeXrmEasy.Integration.{crm version}, which is the next bullet point and would contain nothing but extensions of the real context as suggested by @rmsryu. 

**2) Integration testing**

In relation to the points @daryllabar mentioned, yeah:

1) Define all of your entity Ids that you will be creating as a part of your test ahead of time, so they can automagically be cleaned up pre and post test. => Agree, otherwise using Guid.NewGuid() you'll be always creating new records every single time, yep.

2) In relation to this point and what @nicknow  also said, I can possibly see value whenever you want to test with a bunch of different records....  but having to initialise records in code, does have benefits, specially if you use early bound types because:
 
- You'll get compile errors whenever the schema changes, for example after renaming / deleting an attribute, even before running the tests, as opposed to using the JSON approach, which would be a bit more hard to figure it out when feeding the .Initialize dynamically.

- You just need to setup the records with the minimum attributes required for your unit test, nothing   else. This is handy, because you are testing your code with possibly null values too.

- If there are records shared across tests, you could add extension methods to add those common records to the list too (ex: I am also using FakeXrmEasy for server side unit testing of MVC web applications, where I am constantly injecting a fake identity (logged in contact and account) via extension methods. This has to be implemented once really. Ex:

    .Initialize(new List<Entity>() { 
            record1, record2 
    }.AddFakeIdentities());

    
    
I could see advantages though of using the JSON approach whenever you need to perform complex calculations which require a bunch of records....

Then that would be a 2 step process, as suggested by @nicknow 

   2.1.  Retrieving a bunch of records from CRM, and generate a JSON file for them.
   2.2.  Feed them into the .Initialize().

 I was wondering.... what if 2.1. is actually a VS extension, we could use whenever we want to pull latest data? Like, from visual studio, select a .json file, right click -> Refresh data... or similar .
And then 2.2 could be incorporated into FakeXrmEasy.Integration by just using configuration settings (i.e. whenever .Initialize() is called, used the .json file located in this specific path?)

Finally, in relation to the last point:

_3. Handle any potential issues with existing plugins/workflows that may be manipulating the data. Lets say you are testing a plugin that creates a child record. In your test you assert that only one child record exists, after you run your test. Well, if the plugin is registered against CRM, it will also create the child, and your test will fail, since there will be two child records, rather than one. This normally just involves cleaning up some data after the data-initialization for the plugin, but before the actual plugin event. =>_ Yes, or, you make assertions using a where clause, like if there were X child records before the test, there should be X + 1 child records after the test run. This would save the clean-up but it might still be an issue when running tests in parallel.

**Keep in mind:** in unit testing we get tear down for free, because as soon as the XrmFakedContext is disposed, that's picked up by the garbage collection and everything is cleared. 

For integration testing, this is way more tricky as we are **persisting data**. There could be a million of things failing which would make your tests not tearing down properly (i.e. as simple as an exception in a Delete message... or the really famous Generic SQL Exceptions we all love ;) )

That's why the only way I can think of running repeatable tests in integration testing would be by booting a VM / (or Docker containers ;)) on every test run, so it always starts from a fresh install, and always returns the same result, for the same given tests (hopefully!).

What do you think?

**Comment by daryllabar** [comment]
When it comes with creating unit tests that actually use CRM, for the last 3 years, I've primarily been concerned with:

- Making sure that the code I have locally works when using a real version of CRM to test with, and not just a fake. 

I have not been concerned with:

- setting up a Test Environment with known test data, with known test cases, that are run automatically either on every check-in or on a schedule (i.e. daily)

To that end, the XrmUnitTest framework has tons of features that make my primarily concern a very dry and IMHO minimal code required testing experience.

Due to the horrible speed of CRM (If I have 100 tests that require the creation on average of 4 entities per test, at 200ms for Creates, and 300ms for deletes, assuming single threaded interaction with CRM, that's 3.3 minutes required to run the unit tests, which is unacceptable), I do all my test driven development using a Fake IOrganizationService, then, when I'm ready to check in, I run the same test using an IOrganizationService that is connected to CRM.  If that passes, I build and push to my CRM instance, and then run all my unit tests connected to CRM while I go to the bathroom and get a snack, to ensure that the changes made for the new plugin, or maybe even the new plugin itself, don't break any existing plugin tests.

There is nothing in the XrmUnitTest framework that is dependent on my specific FakeIOrganizationService Implementation.  With that being said, **I would be happy to look into the spending the time porting it over to utilizing the FakeXrmEasy, if it is something that we feel people could get on board with**.

[Here is part of a video I did](https://youtu.be/dRz1X6pdsk4?t=2116) explaining the how I deal with connecting to CRM, and having environment agnostic tests.  So if something is failing in QA, you can just point your unit tests to QA, immediately see what the problem is (maybe a new Attribute or a Security Role Permission didn't get promoted) and get back to creating new functionality, not fighting bugs.  Feel free to stop the video at the 58 minute mark, since that is when I talk about custom workflow activity testing which is almost identical.  

Please let me know if this is something we can all collaborate on together, or if your vision FakeXrmEasy doesn't a-line at all to what I'm describing.

**Comment by daryllabar** [comment]
@nicknow, side note, your comment:

> As to @daryllabar point # 2: One area of work that might be very valuable (and I may, if time permits, take a look at getting it going myself) is creating an enhancement and tool to make it easier to hydrate an instance of XrmFakedContext. Right now it is all coded. What I would like to see is an application that can read data from an existing Dynamics CRM instance, save that data (i.e., a JSON file), and then read them into a List<Entity> object to feed to XrmFakedContext.Initialize.
This exact feature is fully implemented in XrmUnitTest.  You create an assumption class
``` C#
public class BranchIndy : EntityDataAssumptionBaseAttribute
{
    protected override Entity RetrieveEntity(IOrganizationService service)
    {
        return service.GetFirstOrDefault<new_branch>(new_branch.Fields.BranchName, "Indy");
    }
}
```
then decorate any test with the assumption attribute
``` C#
[BranchIndy]
private class BranchClosed_Should_NotBeAbleToScheduleJob : TestMethodClassBase
{
    protected override void Test(IOrganizationService service)
    {
        // At this point, when using a real Crm Connection, it will have already asserted that BranchIndy exists, and seralized the entity to disk for use when running local unit tests.  When using a Fake IOrganizationService, the BranchIndy will already have been deserialized, and created in memory.
    }
}
```

**Comment by jordimontana82** [comment]
@daryllabar Of course, any collaboration would be more than welcome +1.

I agree with your point:

_Making sure that the code I have locally works when using a real version of CRM to test with, and not just a fake._

And for this one:

 _Setting up a Test Environment with known test data, with known test cases, that are run automatically either on every check-in or on a schedule (i.e. daily)_

It becomes a pretty must have requirement when multiple developers work in parallel, or even worse, multiple teams of developers working in parallel, as we do in our case. In that environment CI + CD is kind of essential.

Given said that, the goal of FakeXrmEasy is:

- Make the unit testing phase itself for Dynamics CRM as simple as possible, by having already implemented fakes by default, which would speed up the coding of unit tests themselves, making TDD way more cost-effective.
-Those already mocked / faked messages, specially the query engine, allows you to spot issues much earlier in the development lifecycle, in the unit test phase more exactly, because without this, these would have been discovered later on the integration testing phase otherwise ( or QA or UAT or... if really unlucky, Production).
- It gives another scenario for development which is actually being able to code "offline" without even connecting to CRM. At least for things which require custom code, like plugins / codeactivities / backend development in general.

So the scope of FakeXrmEasy is just the unit testing phase, nothing else. This project is actually for .NET unit testing of CRM, and its sibling project, fake-xrm-easy-js, would be the same concept applied to JS on the client-side.

@rmsryu  was suggesting the same approach you were using to point your tests to a real CRM organisation for QA, or other environments, indeed. I probably didn't express that correctly, so apologies : P 

So I'm happy to add a new XrmRealContext class to the existing package. And then having another one for Integration Testing which could be triggered from a TFS / TeamCity / AppVeyor, whatever build, later on in the build, straight after Unit tests.

**Comment by jordimontana82** [comment]
@rmsryu @daryllabar @craigseymour @nicknow 

Added a few commits with the new XrmRealContext class, which basically overrides a new GetOrganizationService returning a real CRM organisation service based on a CRM connection string.

I implemented it for 2011 but added other paths for 2013, 2015 and 2016 (you'll see precompilation symbols as the codebase is shared across CRM versions).

Now, as I added this from a hotel I couldn't test it really.

Please have a look at let me know what you think.

I'll raise a separate issue for proper "Integration testing" we could use to brainstorm ideas if you want.

**Comment by jordimontana82** [comment]
Separate issue for integration testing: #88

**Comment by daryllabar** [comment]
Do you have any documentation on how to create the connection,  or just a place in code to look at it?

**Comment by jordimontana82** [comment]
Hi @daryllabar 

Sorry no documentation yet.... but, the [here are the bits](https://github.com/jordimontana82/fake-xrm-easy/blob/6f4848a3211126a96462e3b26613f3307feae8ba/FakeXrmEasy.Shared/XrmRealContext.cs#L42-L49) which would setup a connection, based on a connection string you could add in your unit test project's app.config.

More info:
https://msdn.microsoft.com/en-us/library/jj602970.aspx

**Comment by jordimontana82** [comment]
Just pushed a new version 1.16.0 which contains the new XrmRealContext class.

The service needs a connectionString parameter called "fakexrmeasy-connection" to establish the connection (you'll find examples if you look for it in on git).  And it will also bypass the .Initialize() method too.

**Comment by jordimontana82** [comment]
Hi lads

@rmsryu @daryllabar @nicknow @DigitalFlow @ccellar @craigseymour @ccellar 

The change made it to the latest version 1.16.0, would you add something extra in there? Or any other feedback?

**Comment by daryllabar** [comment]
Couple thoughts:

1. **Create Split Config Settings.**  I've found it beneficial to split out these settings in the app.config.  At most of my clients We've had very similar environment names ("Dev.something.com", "QA'something.com") so everything could be exactly the same, except for the environment name.  It was pretty easy to find it and update it, rather than having to look at a huge string.

2. **Password Security** I've also had the concept of reading the settings from a .user.config file that never actually get's checked in, and have some default settings from a .config file.  This helps to alleviate users from checking their password into Source Control.

**Comment by jordimontana82** [comment]
@daryllabar Thx for your feedback!

Indeed! I was hoping to use different configuration settings in VS for 1, and depending on the configuration, use one setting or another using web.config transforms, which depend on a Configuration. Cause then in theory we just need to maintain the 1 connectionString parameter name which will be dynamically changed based on the configuration, and the configuration could be also passed as a parameter from a console runner too (i.e. /Configuration: "Test" will use the web.config transform which points to "QA.something.com").

Totally agree with 2 tho.  How would you do it? Could we add separate settings for credentials which might override the ones specified in the connection string Url maybe?  I'd like the User settings to be an optional configuration. I.e. for some environments, we used shared credentials, in that case, we don't really care much about those credentials for the dev environment. Staging / Production will be a completely different story where the credentials are not known and also encrypted in IIS :)

**Comment by daryllabar** [comment]
You'd need a flag or enum in the config to specify what credentials to use.  For on prem, you wouldn't even want to provide any credentials potentially, and just use Windows credentials.

**Comment by jordimontana82** [comment]
Yep, for non .user.config settings we could just use the connectionString itself, which in theory supports all the different types of authentication. [Example here](https://github.com/jordimontana82/fake-xrm-easy/blob/8394dd95a3cbbfdce057853351291590c2c548a7/FakeXrmEasy.Tests.2016/app.config#L24-L39)

So say we add a configuration setting like key="fakexrmeasy-usercredentials" value="true/false". When true -> After parsing the connectionString, the credentials from the user.config will be used, otherwise, credentials in the connectionString will be used. And those could be default credentials for AD,,or others in the connectionString for IFD, Online, etc...

That should cover any situation potentially. And if you want to use multiple environments, we can use again different project Configuration settings to override the 2 parameters: connectionString + this new parameter.

Will that work?

**Comment by daryllabar** [comment]
I would say the default settings would be to use a fake CRM service, so if you had an automated build server, it wouldn't require an environment to test with post build.

**Comment by jordimontana82** [comment]
Hi guys,

Based on feedback, we merged a pull request with [a new parameter](https://github.com/jordimontana82/fake-xrm-easy/blob/d09e229989b876d25da3d51a36fc3be40bb409e3/FakeXrmEasy.Shared/XrmRealContext.cs#L30-L34) in the constructor of the XrmRealContext where we can pass the connection string.

If no parameter is used, the default constructor will use the default fakexrmeasy-connection from the app.config.

That should cater for the scenario where you want to use different connections for different environments.

Any other thoughs?

**Comment by jordimontana82** [comment]
Hi guys!

Closing this one. Will reopen if there is further feedback!
